### PR TITLE
Feature/1064 refactor max length prop

### PR DESCRIPTION
**Pull request by GCHQ-Developer-847**
## Summary of the changes
Removed `maxLength` prop from text field and made character count display when the `maxCharacters` prop is used instead. Added `hideCharCount` prop to provide the option to visually hide the character count. Updated tests, including removing `maxLength` usage / replacing with `maxCharacters`.

The character count information for screen readers still remains when the visual character count is hidden. I felt that this would be a good idea to make it more obvious to screen reader users that their input is being limited. 

I have also updated the character count screen reader information to read out how many characters are remaining as originally the aria-live region was just reading out "10 slash 25", for example; now it would say "15 characters remaining", which I felt was more user-friendly and also matches the [GOV.UK](https://design-system.service.gov.uk/components/character-count/) behaviour.

## Related issue
#1064 

### General 

- [x] Changes to docs package checked and committed.
- [x] All acceptance criteria reviewed and met. 

### Testing

- [x] Relevant unit tests and visual regression tests added. 
- [x] Visual testing against Figma component specification completed. 
- [x] Playground stories in React Storybook up to date, with any prop changes and additions addressed. 

### Accessibility 

- [x] Accessibility Insights FastPass performed.
- [x] A11y unit test added and yields no issues.
- [x] A11y plug-in on Storybook yields no issues. 
- [x] Manual screen reader testing performed using NVDA and VoiceOver. 
- [x] Manual keyboard testing for keyboard controls and logical focus order. 
- [x] Correct roles used and ARIA attributes used correctly where required. 
- [x] Logical heading structure is maintained, and the HTML elements used for headings can be changed to fit within the wider page structure. 

### Resize/zoom behaviour 

- [x] Page can be zoomed to 400% with no loss of content. 
- [x] Screen magnifier used with no issues. 
- [x] Text resized to 200% with no loss of content.
- [x] Text spacing increased as per the [WCAG 1.4.12 success criterion](https://www.w3.org/TR/WCAG21/#text-spacing) with no loss of content.

### System modes

- [x] Browser setting 'prefers reduced motion' tested. No animations or motion visible whilst this setting is on. 
- [x] Windows High Contrast mode tested with no loss of content.
- [x] System light and dark mode tested with no loss of content.
- [x] Browser support tested (Chrome, Safari, Firefox and Edge). 

### Testing content extremes

- [x] Min/max content examples tested with no loss of content or overflow. 
- [x] All prop combinations work without issue. 
- [x] Tested for FOUC (Flash of Unstyled Content) in both SSR (Server-Side Rendering) and SSG (Static Site Generation) settings.
- [x] Controlled and uncontrolled input components tested.

### CODE REVIEW

File: packages/react/src/stories/ic-text-field.stories.mdx [hunk 1]
```diff
@@ -471,12 +447,18 @@ export const Uncontrolled = () => {
     parameters={{ loki: { skip: true } }}
   >
     <IcTextField
-      max-characters={2}
+      maxCharacters={2}
```
Comment (GCHQ-Developer-299) [inline packages/react/src/stories/ic-text-field.stories.mdx:450]: I'm not sure if the warning behaviour is correct - it's displaying when i'm at the limit of characters, i assumed it'd only happen when i'm *over* the limit
Screenshot 2024-07-05 at 15 24 50

**Review by GCHQ-Developer-299 — DISMISSED** [review]

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Comment by ad9242** [comment]
Should the validation\max length check be triggered on first load. For example if you have a value set that exceeds the maxCharacters? 

I have a long `value` set which exceeds the limit.

Initially this displays as follows:
image

After pressing a key in the input field, the warning is applied:
image


The old behaviour appears to be to trigger the error on load:
image

**Comment by ad9242** [comment]
there no longer seems to be a visual test (with baseline image) where the maxCharacters is exceeded. I think it would be good to have one, just to ensure styling doesn't change

**Comment by GCHQ-Developer-847** [comment]
> Should the validation\max length check be triggered on first load. For example if you have a value set that exceeds the maxCharacters?
> 
> I have a long `value` set which exceeds the limit.
> 
> Initially this displays as follows: image
> 
> After pressing a key in the input field, the warning is applied: image
> 
> The old behaviour appears to be to trigger the error on load: image

With the changes in this PR, it is not possible to get the "maximum length exceeded" error when typing in characters now because it prevents any more input once the maxCharacters amount has been reached.

I have pushed up some changes now which just cut off the value based on the maxCharacters. So now, if the `value` prop is set on page load, or if the prop is updated, the input in the text field cannot be over the maxCharacters amount. I felt that this would be the best solution as it essentially matches the behaviour you were to get if you actually typing in the text field (including the fact that if you try to type more characters, the warning will then appear). Hopefully this is ok :)

**Comment by GCHQ-Developer-847** [comment]
> there no longer seems to be a visual test (with baseline image) where the maxCharacters is exceeded. I think it would be good to have one, just to ensure styling doesn't change

It is no longer possible to go over the maximum allowed number of characters so this error cannot occur any more which is why the image / test has been removed completely

**Comment by ad9242** [comment]
> > Should the validation\max length check be triggered on first load. For example if you have a value set that exceeds the maxCharacters?
> > I have a long `value` set which exceeds the limit.
> > Initially this displays as follows: image
> > After pressing a key in the input field, the warning is applied: image
> > The old behaviour appears to be to trigger the error on load: image
> 
> After the changes in this PR, it is not possible to get the "maximum length exceeded" error when typing in characters now because it prevents any more input once the maxCharacters amount has been reached.
> 
> I have pushed up some changes now which just cut off the value based on the maxCharacters. So now, even if the `value` prop is set on page load, or if the prop is updated, the value displayed in the text field cannot be over the maxCharacters amount. I felt that this would be the best solution as it essentially matches the behaviour you were to get if you actually typing in the text field (including the fact that if you try to type more characters after the prop has been set, the warning will then appear). Hopefully this is ok :)

The only slight issue I can see is that, if the initial value gets truncated, it will trigger an icChange event with the new shortened value. Also, if the data was loaded into a form and then gets submitted, the new shorter value would get submitted and, without some validation being done, potentially overwrite existing data on the server side. I think it would be good to just make it clearer (in the `maxCharacters` prop description?) that the value will get truncated if it exceeds the limit. Just so users are aware and can ensure they validate values

**Comment by GCHQ-Developer-847** [comment]
> > > Should the validation\max length check be triggered on first load. For example if you have a value set that exceeds the maxCharacters?
> > > I have a long `value` set which exceeds the limit.
> > > Initially this displays as follows: image
> > > After pressing a key in the input field, the warning is applied: image
> > > The old behaviour appears to be to trigger the error on load: image
> > 
> > 
> > After the changes in this PR, it is not possible to get the "maximum length exceeded" error when typing in characters now because it prevents any more input once the maxCharacters amount has been reached.
> > I have pushed up some changes now which just cut off the value based on the maxCharacters. So now, even if the `value` prop is set on page load, or if the prop is updated, the value displayed in the text field cannot be over the maxCharacters amount. I felt that this would be the best solution as it essentially matches the behaviour you were to get if you actually typing in the text field (including the fact that if you try to type more characters after the prop has been set, the warning will then appear). Hopefully this is ok :)
> 
> The only slight issue I can see is that, if the initial value gets truncated, it will trigger an icChange event with the new shortened value. Also, if the data was loaded into a form and then gets submitted, the new shorter value would get submitted and, without some validation being done, potentially overwrite existing data on the server side. I think it would be good to just make it clearer (in the `maxCharacters` prop description?) that the value will get truncated if it exceeds the limit. Just so users are aware and can ensure they validate values

Added a note to the prop description 👍

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx [hunk 2]
```diff
@@ -331,16 +339,19 @@ export class TextField {
...
     if (this.value !== this.initialValue) {
       this.watchValueHandler(this.value);
+    } else {
+      if (this.maxCharacters > 0) {
+        this.value = this.value.substring(0, this.maxCharacters);
+      }
     }
 
-    this.getMaxLengthExceeded(this.value);
+    this.getMaxValueExceeded(this.value);
```
Comment (ad9242) [inline packages/web-components/src/components/ic-text-field/ic-text-field.tsx:348]: just realized this is slightly inefficient as both `this.getMaxValueExceeded` and `this.getMaxCharactersReached` may have already been called due to the `watchValueHandler` being called directly (line 341), or indirectly when `this.value` is changed on line 344

**Review by ad9242 — CHANGES_REQUESTED** [review]

File: packages/web-components/src/components/ic-text-field/ic-text-field.tsx (hunk 2 above)
Comment (GCHQ-Developer-847) [reply packages/web-components/src/components/ic-text-field/ic-text-field.tsx:346]: I can't seem to get it to work...

I tried moving those two methods into an `else` on line 344, but then [tests were failing](https://github.com/mi6/ic-ui-kit/actions/runs/9940487837/job/27457312100). It was like line 343 was not causing the `watchValueHandler` to be called at all to set the character count correctly.

So I'm a bit confused 😅 don't know if it's just best to keep it as is even though it is a bit inefficient

**Review by GCHQ-Developer-299 — APPROVED** [review]

**Pull request merged**